Mathlib/NumberTheory/Cyclotomic/CyclotomicCharacter.lean

```diff
@@ -0,0 +1,201 @@
+    ZMod.val_int_cast, ← Subgroup.coe_zpow]
+  exact Units.ext_iff.1 <| SetCoe.ext_iff.2 <| Group.pow_eq_zpow_mod _ pow_card_eq_one
+
+theorem spec' (g : L ≃+* L) {n : ℕ+} {t : Lˣ} (ht : t ∈ rootsOfUnity n L) :
+    g t = t ^ (χ n g).val :=
+  spec g ⟨t, ht⟩
+
+theorem spec'' (g : L ≃+* L) {n : ℕ+} {t : L} (ht : IsPrimitiveRoot t n) :
+    g t = t ^ (χ n g).val :=
+  spec' g (SetLike.coe_mem ht.toRootsOfUnity)
+
+-- this is in the wrong place I guess
+lemma ext {G : Type _} [Group G] [Fintype G] [IsCyclic G]
+    {d : ℕ} {a b : ZMod d} (hGcard : Fintype.card G = d) (h : ∀ t : G, t^a.val = t^b.val) :
+  a = b := by
+  obtain ⟨g, hg⟩ := IsCyclic.exists_generator (α := G)
+  have hgord := orderOf_eq_card_of_forall_mem_zpowers hg
+  specialize h g
+  rw [pow_eq_pow_iff_modEq, hgord, hGcard, ← ZMod.nat_cast_eq_nat_cast_iff] at h
+  subst hGcard
+  simpa [ZMod.nat_cast_val, ZMod.cast_id'] using h
+
+lemma id : χ n (RingEquiv.refl L) = 1 := by
+  refine ext (G := rootsOfUnity n L) rfl ?_
+  intro ζ
+  ext
+  rw [Subgroup.coe_pow, ← spec]
+  have : 1 ≤ Fintype.card { x // x ∈ rootsOfUnity n L } := Fin.size_positive'
+  obtain (h | h) := this.lt_or_eq
+  · have := Fact.mk h
+    simp [ZMod.val_one]
+  · have := Fintype.card_le_one_iff_subsingleton.mp h.ge
+    obtain rfl : ζ = 1 := Subsingleton.elim ζ 1
+    simp
+
+lemma comp (g h : L ≃+* L) : χ n (g * h) =
+    χ n g * χ n h := by
+  refine ext (G := rootsOfUnity n L) rfl ?_
+  intro ζ
+  ext
+  rw [Subgroup.coe_pow, ← spec]
+  change g (h (ζ : Lˣ)) = _
+  rw [spec, ← Subgroup.coe_pow, spec, mul_comm, Subgroup.coe_pow, ← pow_mul, ← Subgroup.coe_pow]
+  congr 2
+  simp only [pow_eq_pow_iff_modEq, ← ZMod.nat_cast_eq_nat_cast_iff, SubmonoidClass.coe_pow,
+    ZMod.nat_cast_val, Nat.cast_mul, ZMod.cast_mul (m := orderOf ζ) orderOf_dvd_card]
+
+end ModularCyclotomicCharacter
+
+variable (L)
+
+/-- Given a positive integer `n`, `ModularCyclotomicCharacter' n` is a
+multiplicative homomorphism from the automorphisms of a field `L` to `ℤ/dℤˣ`,
+where `d` is the number of `n`'th roots of unity in `L`. It is uniquely
+characterised by the property that `g(ζ)=ζ^(ModularCyclotomicCharacter n g)`
+for `g` an automorphism of `L` and `ζ` an `n`th root of unity. -/
+noncomputable
+def ModularCyclotomicCharacter' (n : ℕ+) :
+    (L ≃+* L) →* (ZMod (Fintype.card { x // x ∈ rootsOfUnity n L }))ˣ := MonoidHom.toHomUnits
+  { toFun := ModularCyclotomicCharacter.toFun n
+    map_one' := ModularCyclotomicCharacter.id n
+    map_mul' := ModularCyclotomicCharacter.comp n }
+
+/-- Given a positive integer `n` and a field `L` containing `n` `n`th roots
+of unity, `ModularCyclotomicCharacter n` is a multiplicative homomorphism from the
+automorphisms of `L` to `ℤ/nℤ`. It is uniquely characterised by the property that
+`g(ζ)=ζ^(ModularCyclotomicCharacter n g)` for `g` an automorphism of `L` and `ζ` any `n`th root
+of unity. -/
+noncomputable def ModularCyclotomicCharacter (n : ℕ+)
```

Comment:
`n` could be `Nat` and implicit here (the cardinality in `hn` cannot be zero).

Reply:
Unfortunately `rootsOfUnity n` needs `n : ℕ+` so I don't see how to make this work.